cranelift/docs/regalloc2-integration.md

```diff
@@ -0,0 +1,245 @@
+  insert allocations, then finally in `emit`). This is actually pretty
+  important, as the match is very control-flow intensive and hard to
+  predict.
+
+- Debug info is handled natively by the regalloc now, rather than
+  reconstructed from an awkward post-pass. regalloc.rs did not have a
+  way of tracking where a value ends up based on arbitrary labels, so
+  we inserted pseudo-instructions ("value markers") that consumed
+  program-level values, let those go through regalloc, then did a
+  dataflow analysis on the machine code with code offsets as locations
+  (!) to build debug info. This was awkward, slow, and
+  error-prone. regalloc2 now has a mechanism to apply labels to VRegs
+  for certain ranges, and will indicate which Allocations have those
+  labels on the output side.
+  
+## x86-64 backend effects
+
+- Instructions really become three-operand, including during
+  pretty-print. This is kind of awkward, but necessary to show
+  pre-regalloc vcode correctly. Previously, many of the instructions
+  (e.g., `AluRmiR` and similar) had been made three-operand (or more
+  precisely, to have separate sources and a dest) as part of the ISLE
+  migration, but pretty-printing showed them in conventional
+  two-operand x86 form.
+  
+  Pretty-printing takes an `&[Allocation]` as well and uses the same
+  `AllocationConsumer` to print allocated registers when producing the
+  post-regalloc disassembly.
+
+- As a consequence of reused-input operands, we can specify the
+  constraint that (e.g.) `src1 == dst` by emission time directly to
+  regalloc2, and get the same allocation for both, so we can emit an
+  actual x86 two-operand instruction. This completely replaces the
+  separate-move idiom, mostly generated by mov-mitosis in today's
+  backend. This ends up being both more efficient (we start with many
+  fewer instructions and the regalloc inserts moves only where
+  necessary, rather than the other way around relying on move elision)
+  and cleaner to read in pre-regalloc disassemblies.
+
+- Can still generate non-SSA code and use mod ("modify") regalloc
+  operands when necessary, and can still use fixed RealRegs.
+  
+  Aside: we should work to reduce these over time, and ideally we
+  should eventually remove mod operands and non-SSA support from
+  regalloc2; this would allow additional optimizations. (We can, for
+  example, handle spilling a little more efficiently by skipping if we
+  know a reloaded vreg hasn't changed with a new def. The
+  redundant-move eliminator in regalloc2 kind of does this today, but
+  not perfectly.)
+
+## other backends
+
+- This is still in progress. Aarch64 should be relatively easy: most
+  of its instructions are already in three-operand form. It's a mostly
+  mechanical transformation of the "instruction library". S390x is
+  possibly a bit more involved, but now that all design decisions are
+  settled, I expect these both together to take a few days at most.
```

Comment:
This is out of date now. Not sure whether we want to keep this whole doc around or not?

Reply:
Yeah, it was mostly just to help the reviewers (ie, you) as a delta, rather than a snapshot of the whole design, so I'll go ahead and delete it I think. It'll still be around in my branch (I'll delete it in a separate commit) if anyone wants to dig it up later.